src/composer.ts

```diff
@@ -53,51 +99,51 @@ export class Composer<TContext extends Context>
   /**
    * Registers middleware for handling provided update types.
    */
-  on(
-    updateTypes: MaybeArray<tt.UpdateType | tt.MessageSubTypes>,
-    ...fns: NonemptyReadonlyArray<Middleware<TContext>>
+  on<U extends tt.UpdateType | tt.MessageSubType>(
+    updateType: MaybeArray<U>,
+    ...fns: NonemptyReadonlyArray<Middleware<MatchedContext<TContext, U>>>
```

Comment:
I'd love that! It's not really in the scope of this PR, though, is it? `router.ts`, `types.ts`, and everything in the scenes directory are so far unaffected by this PR, so it may make sense to do a refactoring PR

Reply:
> It's not really in the scope of this PR, though, is it?

This PR already does way more than leveraging unionization, so it's fine.

`scenes/**` are already changed in #1194, the other files can be changed while doing some other changes to them, don't touch them here.

Reply:
> This PR already does way more than leveraging unionization

What would that be?

I mean, I see how it is convenient to do the renaming here given that the code is modified anyway. But what changes are so far not required when we want to make use of unionization?

Reply:
`MatchedMiddlewaware`, removal of `Parameters`, explicit return type annotations...

Reply:
> don't touch them here

...

Reply:
I'll do a partial revert, hang on

Reply:
Sorry for not being clear enough

Reply:
> `MatchedMiddlewaware`

necessary to do this while staying DRY

> removal of `Parameters`

necessary because the type variable now occurs in the function arguments, and we cannot do something like `typeof Composer[<C>'url']`, so we have to reintroduce the type variable to have access to it

> explicit return type annotations

necessary because it makes type inference much faster (I was basically just looking for places where we could precompute types as much as possible)